Framework/PythonInterface/plugins/algorithms/Abins.py

```diff
@@ -286,16 +315,65 @@ def _create_workspaces(self, atoms_symbols=None, s_data=None):
         for s in masses:
             masses[s] = sorted(list(set(masses[s])))
 
-        for symbol in atoms_symbols:
-
-            sub = len(masses[symbol]) > one_m or abs(Atom(symbol=symbol).mass - masses[symbol][0]) > eps
-            for m in masses[symbol]:
-                result.extend(self._atom_type_s(num_atoms=num_atoms, mass=m, s_data_extracted=s_data_extracted,
-                                                element_symbol=symbol, temp_s_atom_data=temp_s_atom_data,
-                                                s_atom_data=s_atom_data, substitution=sub))
+        result = []
 
+        if atoms_symbols is not None:
+            for symbol in atoms_symbols:
+                sub = len(masses[symbol]) > one_m or abs(Atom(symbol=symbol).mass - masses[symbol][0]) > eps
+                for m in masses[symbol]:
+                    result.extend(self._atom_type_s(num_atoms=num_atoms, mass=m, s_data_extracted=s_data_extracted,
+                                                    element_symbol=symbol, temp_s_atom_data=temp_s_atom_data,
+                                                    s_atom_data=s_atom_data, substitution=sub))
```

Comment:
Just a suggestion, maybe the above lines of code could be in a separate function?

Reply:
I'll have a tinker.

Reply:
I've had a little play around and I don't think it is helpful. It's going to need a big messy argument list anyway, the benefit would be to avoid seeing it twice.
This looping/reasoning is the main purpose of the _create_workspaces function so I think it's best to keep the loops and if statements visible so they are easier to reason about.

What I can do to clean this up is move the "masses" table building logic into its own function as that's a bit more self-contained. I think that does improve the readability and makes it clearer that these loops are the "meat" of the function and the stuff above is setup.